src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs

```diff
@@ -0,0 +1,355 @@
+                                    TokenList(Token(SyntaxKind.InternalKeyword), Token(SyntaxKind.PartialKeyword)),
+                                    SyntaxKind.InterfaceDeclaration,
+                                    Identifier("Native"),
+                                    null);
+
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            // Get all methods with the [GeneratedComInterface] attribute.
+            var attributedInterfaces = context.SyntaxProvider
+                .ForAttributeWithMetadataName(
+                    TypeNames.GeneratedComInterfaceAttribute,
+                    static (node, ct) => node is InterfaceDeclarationSyntax,
+                    static (context, ct) => context.TargetSymbol is INamedTypeSymbol interfaceSymbol
+                        ? new { Syntax = (InterfaceDeclarationSyntax)context.TargetNode, Symbol = interfaceSymbol }
+                        : null)
+                .Where(
+                    static modelData => modelData is not null);
+
+            var interfacesWithDiagnostics = attributedInterfaces.Select(static (data, ct) =>
+            {
+                Diagnostic? diagnostic = GetDiagnosticIfInvalidTypeForGeneration(data.Syntax, data.Symbol);
+                return new { data.Syntax, data.Symbol, Diagnostic = diagnostic };
+            });
+
+            // Split the methods we want to generate and the ones we don't into two separate groups.
+            var interfacesToGenerate = interfacesWithDiagnostics.Where(static data => data.Diagnostic is null);
+            var invalidTypeDiagnostics = interfacesWithDiagnostics.Where(static data => data.Diagnostic is not null);
+
+            var interfaceContexts = interfacesToGenerate.Select((data, ct) =>
+            {
+                // Start at offset 3 as 0-2 are IUnknown.
+                // TODO: Calculate starting offset based on base types.
+                // TODO: Extract IID from source.
+                return new ComInterfaceContext(3, Guid.Empty);
+            });
+
+            context.RegisterSourceOutput(invalidTypeDiagnostics, static (context, invalidMethod) =>
+            {
+                context.ReportDiagnostic(invalidMethod.Diagnostic);
+            });
+
+            // Zip the incremental interface context back with the symbols and syntax for the interface
+            // to calculate the methods to generate.
+            // The generator infrastructure preserves ordering of the tables once Select statements are in use,
+            // so we can rely on the order matching here.
+            var methodsWithDiagnostics = interfacesToGenerate
+                .Zip(interfaceContexts)
+                .SelectMany(static (data, ct) =>
+            {
+                ContainingSyntaxContext containingSyntax = new(data.Left.Syntax);
+                Location interfaceLocation = data.Left.Syntax.GetLocation();
```

Comment:
This left/right logic is really confusing to review. I'm not convinced it is any better than arbitrary unions in C - especially when trying to review outside of VS. What do you think of extracting them at the top of this method?

```csharp
<TYPE_1> l = data.Left;
<TYPE_2> r = data.Right;
```

Is this helpful in practice?

Reply:
I'll deconstruct the tuple into named variables at the top. I agree that it will improve readability